src/glb_writer.rs

```diff
@@ -0,0 +1,561 @@
+use std::io::prelude::*;
+
+use crate::point::OutputPoint;
+
+/// glTF version number. 2.0 is the latest as of this writing.
+const GLTF_VERSION: u32 = 2;
+/// glTF mode for point clouds
+const GLTF_POINTS: u32 = 0;
+/// glTF constant for FLOAT component type
+const GLTF_FLOAT: u32 = 5126;
+/// glTF constant for UNSIGNED_BYTE component type
+const GLTF_UNSIGNED_BYTE: u32 = 5121;
+/// Length of the glTF header
+const GLTF_HEADER_LENGTH: u32 = 12;
+/// Length of a chunk header (length + type)
+const GLTF_CHUNK_HEADER_LENGTH: u32 = 8;
+
+/// Size of a vec3
+const SIZE_VEC3: u32 = 3 * 4;
+/// Size of an RGB color encoded as unsigned bytes
+const SIZE_COLOR_RGB: u32 = 3;
+// Size of an unsigned 64-bit integer in bytes
+const SIZE_U64: u32 = 8;
+// Size of an unsigned 16-bit integer in bytes
+const SIZE_U16: u32 = 2;
+// Size of an unsigned 8-bit integer in bytes
+const SIZE_U8: u32 = 1;
+
+/// Alignment for each buffer view. I'm setting them all to 8
+/// as this is the simplest way to satisfy both glTF and EXT_mesh_features
+/// requirements without making the code complex
+const ALIGNMENT: u32 = 8;
+
+const PADDING_BINARY: u8 = 0x00;
+const PADDING_JSON: u8 = ' ' as u8;
+
+struct BufferView {
+    byte_offset: u32,
+    byte_length: u32,
+    padding_length: u32,
+}
+
+impl BufferView {
+    pub fn new() -> Self {
+        Self {
+            byte_offset: 0,
+            byte_length: 0,
+            padding_length: 0
+        }
+    }
+
+    // Compute the offset after this buffer view and after the pad
+    pub fn after_offset(&self) -> u32 {
+        self.byte_offset + self.byte_length + self.padding_length
+    }
+}
+
+struct Accessor {
+    buffer_view: u32,
+    component_type: u32,
+    /// minimum value of each component of the accessor. Only required for
+    /// position
+    min: Option<Vec<f32>>,
+    /// minimum value of each component of the accessor. Only required for
+    /// position
+    max: Option<Vec<f32>>,
+}
+
+impl Accessor {
+    pub fn new() -> Self {
+        Self {
+            buffer_view: 0,
+            component_type: 0,
+            min: None,
+            max: None
+        }
+    }
+}
+
+struct Chunk {
+    chunk_length: u32,
+    padding_length: u32
+}
+
+impl Chunk {
+    pub fn new() -> Self {
+        Self {
+            chunk_length: 0,
+            padding_length: 0,
+        }
+    }
+
+    pub fn data_length(&self) -> u32 {
+        self.chunk_length + self.padding_length
+    }
+
+    pub fn total_length(&self) -> u32 {
+        GLTF_CHUNK_HEADER_LENGTH + self.chunk_length + self.padding_length
+    }
+}
+
+pub struct GlbWriter {
+    /// Number of points in this glTF point cloud
+    point_count: u32,
+    /// Total length of binary glTF, including header and all chunks
+    total_length: u32,
+    json_chunk: Chunk,
+    binary_chunk: Chunk,
+    buffer_length: u32,
+    accessor_position: Accessor,
+    accessor_color: Accessor,
+    accessor_feature_ids: Accessor,
+    buffer_view_position: BufferView,
```

Comment:
I don't like how hard-coded this structure is. when using the `chaos_game` algorithm, some of the metadata values are irrelevant (e.g. point_id is always 0, the accessor feature IDs is irrelevant). I should try refactoring this to take a vector of BufferViews. If needed, the BufferView could store some of the JSON

Reply:
Eh, there's better things I could be doing with my time here. I'll leave it as is.